internal/controllers/machine/machine_controller_phases.go

```diff
@@ -312,3 +312,44 @@ func (r *Reconciler) reconcileInfrastructure(ctx context.Context, cluster *clust
 	m.Spec.ProviderID = pointer.StringPtr(providerID)
 	return ctrl.Result{}, nil
 }
+
+func (r *Reconciler) reconcileCertificateExpiry(ctx context.Context, _ *clusterv1.Cluster, m *clusterv1.Machine) (ctrl.Result, error) {
+	var annotations map[string]string
+
+	if !util.IsControlPlaneMachine(m) {
+		// If the machine is not a control plane machine, return early.
+		return ctrl.Result{}, nil
+	}
+
+	if m.Spec.Bootstrap.ConfigRef != nil {
+		bootstrapConfig, err := external.Get(ctx, r.Client, m.Spec.Bootstrap.ConfigRef, m.Namespace)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// Check for certificate expiry information in the bootstrap config.
+		annotations = bootstrapConfig.GetAnnotations()
+		if expiry, ok := annotations[clusterv1.MachineCertificatesExpiryDateAnnotation]; ok {
+			expiryTime, err := time.Parse(time.RFC3339, expiry)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to parse expiry date")
```

Comment:
let's mention the resource where we are reading the annotation from, e.g. ~ `"failed to parse expiry date from annotation on %s", klog.KObj...)`
same below in l.348